### PR TITLE
util: Duration Add

### PR DESCRIPTION
**Pull request by hawkingrei**
for the addtime function
https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_addtime

### CODE REVIEW

File: util/types/time.go [hunk 1]
```diff
@@ -673,6 +673,14 @@ type Duration struct {
 	Fsp int
 }
 
+//Add adds d to d, returns a duration value.
+func (d Duration) Add(v *Duration) Duration {
```
Comment (shenli) [inline util/types/time.go]: Should we consider v == null?

Comment (hawkingrei) [inline util/types/time.go]: Thank you , fix it

**Comment by shenli** [comment]
LGTM
@tiancaiamao @XuHuaiyu PTAL

File: util/types/time.go (hunk 1 above)
Comment (tiancaiamao) [inline util/types/time.go]: Why not `Duration` as parameter?

File: util/types/time.go [hunk 2]
```diff
@@ -673,6 +673,17 @@ type Duration struct {
...
+	if v == nil {
+		return d
+	}
+	if d.Fsp >= v.Fsp {
+		return Duration{Duration: d.Duration + v.Duration, Fsp: d.Fsp}
```
Comment (tiancaiamao) [inline util/types/time.go]: Will `+` operation overflow?

Comment (hawkingrei) [inline util/types/time.go]: it will overflow. I will fix it

File: util/types/time.go (hunk 1 above)
Comment (hawkingrei) [inline util/types/time.go]: I reference the code  
https://github.com/pingcap/tidb/blob/master/util/types/time.go#L464

File: util/types/time.go (hunk 2 above)
Comment (hawkingrei) [inline util/types/time.go]: thx ,fixed @tiancaiamao

**Comment by hawkingrei** [comment]
Why not Duration as parameter?

 - keey code style

func (t *Time) Add(d *Duration) Duration 
https://github.com/pingcap/tidb/blob/master/util/types/time.go#L464

**Comment by hawkingrei** [comment]
@tiancaiamao  PTAL

**Comment by tiancaiamao** [comment]
I think it's better to change 
`func (t *Time) Add(d *Duration) Duration`
 to 
`func (t *Time) Add(d Duration) Duration`

As you see, `Duration` is used everywhere.
The return value of Add is `Duration`, 
Method is defined on `Duration` rather than `*Duration` 
Datum in datum.go also use Duration ...

`func (t *Time) Add(d *Duration) Duration` must be a mistake!

**Comment by hawkingrei** [comment]
fixed it for
func (d Duration) Add(d Duration) Duration

for Add of time,i will create rp to fix it

@tiancaiamao  PTAL

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**